EL-3660 - typeahead dynamic dropdown

TaraTodd123 opened the pull request:
#### Checklist
<!-- Use an 'x' to check those which apply. -->
* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-3660

#### Description of Proposed Changes

- added a dropDirection 'auto' setting
- Added tests e2e/karma for the typeahead and the dropDirection settings
- Added screenshot tests for typeahead.
- Created an injectable service for the functionality of checking the component in the viewport and how close it is to the bottom of the window.
- Updated the documentation, examples and plunker for the typeahead component and select component.

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-3600-typeahead-dynamic-dropdown

#### Downstream Build(s)

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/caf_CI_ux-aspects-micro-focus_EL-3600-typeahead-dynamic-dropdown

buildmachine-sou-jenkins2 commented:
The Documentation QA site for this branch has been built:
https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-3600-typeahead-dynamic-dropdown

TaraTodd123 commented:
Moved functionality into an injectable service. Moved some of the tests to karma tests but I have had to keep the dropDirection auto functionality in a e2e test as due to drop direction changing when on scrolling to the bottom of the page or being near the bottom of the page I could not get this functionality to show in a karma test.

donna-marijne commented on src/services/viewport-listerner/viewport-listener.module.ts:
```diff
@@ -0,0 +1,9 @@
+import { NgModule } from '@angular/core';
+import { ViewportListenerService } from './viewport-listener.service';
+import { ResizeModule } from '../../../dist/directives/resize';
```
I think this is the problem, you've accidentally imported from the dist directory.

TaraTodd123 replied:
Thanks!

donna-marijne commented on e2e/tests/components/typeahead/typeahead.e2e-spec.ts:
```diff
@@ -0,0 +1,27 @@
+
+        await page.typeaheadInput.click();
+        expect(await page.getTypeaheadOptionListClass()).toContain('ux-typeahead-all-options');
+        expect(page.getTypeaheadClass()).not.toContain('drop-up');
```
Should we perhaps include screenshots for the two dropdown positions?

donna-marijne commented on src/services/viewport-listerner/viewport-listener.service.ts:
```diff
@@ -0,0 +1,77 @@
+import { fromEvent, Subject } from 'rxjs';
+
+@Injectable()
+/** Rename  */
```
Possibly something like `PopoverOrientationService` which would indicate its usage?

donna-marijne commented on src/services/viewport-listerner/viewport-listener.service.ts:
```diff
@@ -0,0 +1,77 @@
+export class ViewportListener implements OnDestroy {
+
+    /** Allow subscribing to state changes */
+    change$ = new Subject<ViewportDirection>();
```
This might be better named `orientation$`.

donna-marijne commented on src/services/viewport-listerner/viewport-listener.service.ts:
```diff
@@ -0,0 +1,77 @@
+
+    }
+
+    ngOnDestroy(): void {
```
Looking at the current code, I don't think this will be invoked. `ngOnDestroy` would be called if defined on the service, but not a plain object like this. You could rename this to `destroy` and call it in the `ngOnDestroy` hook of the typeahead component. However, since the code is not really making use of the `ViewportListener` object, you could just move the implementation into the service and eliminate `ViewportListener`. In that case `ngOnDestroy` would be called automatically when the component that injected it gets destroyed.

donna-marijne commented on src/services/viewport-listerner/viewport-listener.service.ts:
```diff
@@ -0,0 +1,77 @@
+    }
+}
+
+export const enum ViewportDirection {
```
Type name may need to update in line with the other changes.

TaraTodd123 commented:
This is ready for review both branches have built with my changes, i've just updated the branch with changes from the develop branch

donna-marijne commented on src/components/typeahead/typeahead.spec.ts:
```diff
@@ -0,0 +1,143 @@
+
+        /* Adding values to typeahead list */
+        for (let index = 0; index < 200; index++) {
+            this.values = [...this.values, chance.name()];
```
I know this doesn't affect the tests as they currently stand, but we should never use chance in a testing scenario. If you could replace it with something deterministic here that'd be great.

TaraTodd123 commented:
Updates made for above

donna-marijne commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -255,6 +255,7 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
     ngOnDestroy(): void {
         this._onDestroy.next();
         this._onDestroy.complete();
+        this._popoverOrientation.createPopoverOrientationListener(this.typeaheadElement.nativeElement, this.typeaheadElement.nativeElement.parentElement).destroy();
```
You'll need to save the original instance as a private property and destroy that, this is just creating a second set of resources and cleaning those up.

TaraTodd123 commented:
Update made.

donna-marijne commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -138,6 +138,8 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
 
     private _onDestroy = new Subject<void>();
 
+    private _popoverOrientationDirection: PopoverOrientationListener;
```
Presumably this should be named `_popoverOrientationListener` now.

donna-marijne commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -144,7 +149,11 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
         public typeaheadElement: ElementRef,
         private _changeDetector: ChangeDetectorRef,
-        private _service: TypeaheadService
+        private _popoverOrientation: PopoverOrientationService,
```
With the change to store the object, this no longer needs to be persisted, so you can remove the `private` and the underscore.

donna-marijne commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -174,6 +183,17 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
+
+        this._popoverOrientationDirection.orientation$.pipe(takeUntil(this._onDestroy))
+            .subscribe(direction => {
+                if (this.dropDirection === 'auto' && direction === PopoverOrientationDirection.Up) {
```
Just noticed that dropDirection = 'up' no longer works, I think this code will need to change. I'd suggest switching away from using renderer and instead computing a property with a hostbinding, e.g.:
```ts
@HostBinding('class.drop-up')
private _dropUp: boolean;
```

You can then update the property from both sources, e.g. set up ngOnChanges to watch `dropDirection`:
```ts
if (changes.dropDirection.currentValue === 'auto') {
  this._dropUp = this._popoverOrientationDirection.orientation$.getValue() === PopoverOrientationDirection.Up;
} else {
  this._dropUp = changes.dropDirection.currentValue === 'up';
}
```

This subscription should then only update `_dropUp` if `dropDirection` is auto.

TaraTodd123 replied:
I tried doing the above but because I am using orientation$ is a Subject I don't have access to valueOf() and some of the times the current value was coming back as undefined. I've pushed a fix that has fixed the 'up' dropDirection.

donna-marijne replied:
You can convert orientation$ to a BehaviorSubject to resolve that.

donna-marijne commented on src/services/popover-orientation/popover-orientation.service.ts:
```diff
@@ -0,0 +1,76 @@
+    }
+}
+
+export const enum PopoverOrientationDirection {
```
Name got a bit messed up, should probably be `PopoverOrientation`?

TaraTodd123 commented:
I think adding an e2e test with a screenshot will help with making sure the dropDirection 'up' and 'down' works as per our functionality is in the docs. As I already have karma tests in for it as well, and they are passing or failing when I say expect the class list to contain the drop-up class and vice versa.

donna-marijne commented on src/services/popover-orientation/popover-orientation.service.ts:
```diff
@@ -28,7 +28,7 @@ export class PopoverOrientationService {
 
     /** Allow subscribing to state changes */
-    orientation$ = new Subject<PopoverOrientation>();
+    orientation$ = new BehaviorSubject<PopoverOrientation>(null);
```
Default orientation is down, so that would make a good initial value here.

TaraTodd123 replied:
updated to down as the initial value

donna-marijne commented on src/services/popover-orientation/popover-orientation.service.ts:
```diff
@@ -1,8 +1,8 @@
-import { ElementRef, Injectable, OnDestroy, Renderer2 } from '@angular/core';
+import { ElementRef, Injectable, Renderer2 } from '@angular/core';
```
Also I think Renderer2 is not being used for anything in this file.

TaraTodd123 replied:
removed Renderer2

TaraTodd123 commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -187,10 +187,10 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
+                    this._dropUp = direction === PopoverOrientation.Up;
                 } else {
-                    this._renderer.removeClass(this.typeaheadElement.nativeElement, 'drop-up');
+                    this._dropUp = this.dropDirection === 'up';
```
removed

TaraTodd123 commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -271,6 +279,9 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
     }
 
+    @HostBinding('class.drop-up')
+    _dropUp: boolean;
```
Its not liking the private flag but ive moved it up to above the constructor

TaraTodd123 commented on src/services/popover-orientation/popover-orientation.service.ts:
```diff
@@ -1,8 +1,8 @@
 import { ResizeService } from '../../directives/resize/index';
 import { takeUntil } from 'rxjs/operators';
-import { fromEvent, Subject } from 'rxjs';
+import { BehaviorSubject, fromEvent, Subject } from 'rxjs';
```
Subject needs to remain for the _onDestroy

TaraTodd123 commented:
That's all the changes up to date now. Just waiting on the build to pass.

donna-marijne commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -144,7 +152,11 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
         public typeaheadElement: ElementRef,
         private _changeDetector: ChangeDetectorRef,
-        private _service: TypeaheadService
+        public popoverOrientation: PopoverOrientationService,
```
This should not be public, it is only used in the constructor body.

TaraTodd123 commented:
Build has finally passed!!
